zhusuan/distributions/multivariate.py

```diff
@@ -29,6 +33,158 @@
+            dtype=dtype,
+            param_dtype=dtype,
+            is_continuous=True,
+            is_reparameterized=is_reparameterized,
+            use_path_derivative=use_path_derivative,
+            group_ndims=group_ndims,
+            **kwargs)
+
+    @property
+    def mean(self):
+        """The mean of the normal distribution."""
+        return self._mean
+
+    @property
+    def cov_tril(self):
+        """
+        The lower triangular matrix in the cholosky decomposition of the
+        covariance.
+        """
+        return self._cov_tril
+
+    def _value_shape(self):
+        return tf.convert_to_tensor([self._n_dim], tf.int32)
+
+    def _get_value_shape(self):
+        if isinstance(self._n_dim, int):
+            return tf.TensorShape([self._n_dim])
+        return tf.TensorShape([None])
+
+    def _batch_shape(self):
+        return tf.shape(self.mean)[:-1]
+
+    def _get_batch_shape(self):
+        if self.mean.get_shape():
+            return self.mean.get_shape()[:-1]
+        return tf.TensorShape(None)
+
+    def _sample(self, n_samples):
+        mean, cov_tril = self.mean, self.cov_tril
+        if not self.is_reparameterized:
+            mean = tf.stop_gradient(mean)
+            cov_tril = tf.stop_gradient(cov_tril)
+
+        def tile(t):
+            new_shape = tf.concat([[n_samples], tf.ones_like(tf.shape(t))], 0)
+            return tf.tile(tf.expand_dims(t, 0), new_shape)
+
+        batch_mean = tile(mean)
+        batch_cov = tile(cov_tril)
+        # n_dim -> n_dim x 1 for matmul
+        batch_mean = tf.expand_dims(batch_mean, -1)
+        noise = tf.random_normal(tf.shape(batch_mean), dtype=self.dtype)
+        samples = tf.matmul(batch_cov, noise) + batch_mean
+        samples = tf.squeeze(samples, -1)
+        # Update static shape
+        static_n_samples = n_samples if isinstance(n_samples, int) else None
+        if self.get_batch_shape():
+            samples.set_shape(tf.TensorShape([static_n_samples])
+                              .concatenate(self.get_batch_shape())
+                              .concatenate(self.get_value_shape()))
+        return samples
+
+    def _log_prob(self, given):
+        mean, cov_tril = (self.path_param(self.mean),
+                          self.path_param(self.cov_tril))
+        log_det = tf.reduce_sum(tf.log(tf.matrix_diag_part(cov_tril)), axis=-1)
+        N = tf.cast(self._n_dim, self.dtype)
+        logZ = - N / 2 * (tf.log(2 * tf.constant(np.pi, dtype=self.dtype)) +
+                          log_det)
+        # logZ.shape == batch_shape
+        if self._check_numerics:
+            logZ = tf.check_numerics(logZ, "log[det(Cov)]")
+        # (given-mean)' Sigma^{-1} (given-mean) =
+        # (g-m)' L^{-T} L^{-1} (g-m) = |x|^2, where Lx = g-m =: y.
+        y = tf.expand_dims(given - mean, -1)
+        L, _ = maybe_explicit_broadcast(
+                cov_tril, y, 'MVNTriL.cov_tril', 'expand_dims(given, -1)')
```

Comment:
MVNTriL to full name and there seems no need to broadcast if we have forced mean and cov_tril to have the same batch_shape in `__init__`.

Reply:
We need this because `given` can have another dimension at head, and `matrix_triangular_solve` doesn't do implicit broadcast.

Reply:
I see. Maybe we should also assert the rank of given to be at least 1.

Reply:
I guess it's covered in [`_check_input_shape`](https://github.com/thu-ml/zhusuan/blob/c2c0287deb8952180ec793dd9c57f5c7453f439e/zhusuan/distributions/base.py#L273) invoked by `Distribution.log_prob`, as MVN.`value_shape` is at least 1.